### PR TITLE
Metrics distribution

### PR DESCRIPTION
**Pull request by MikailDuranQM**
Comparator.compare now offers to compute metrics on masked data or on all data.
ImputerOracle implemented.
Firsts unit tests on imputers.

### CODE REVIEW

File: qolmat/imputations/imputers.py [hunk 1]
```diff
@@ -203,8 +228,14 @@ class ImputerShuffle(Imputer):
     >>> imputor.fit_transform(df)
     """
 
-    def __init__(self, groups: List[str] = []) -> None:
+    def __init__(self, groups: List[str] = [], random_state: int = None) -> None:
         super().__init__(groups=groups, columnwise=True)
+        self.random_state = random_state
+
+    def fit_transform(self, df: pd.DataFrame) -> pd.DataFrame:
+        if isinstance(self.random_state, int):
```
Comment (JulienRoussel77) [inline qolmat/imputations/imputers.py:236]: Cette fonction a l'air de faire ce que tu veux :
https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_random_state.html
Tu peux l'appeler dans le fit_transform de la classe Imputer pour que ça profite à tous les imputeurs stochastiques, en particulier l'EM

File: qolmat/imputations/imputers.py [hunk 2]
```diff
@@ -282,7 +313,7 @@ def fit_transform_element(self, df: pd.DataFrame) -> pd.DataFrame:
         """
         df_out = df.copy()
         for col in df:
-            df_out[col] = pd.Series.shift(df[col], 1).bfill().ffill()
+            df_out[col] = pd.Series.shift(df[col], -1).bfill().ffill()
```
Comment (JulienRoussel77) [inline qolmat/imputations/imputers.py:316]: df_out[col] = df_out[col].bfill()

File: qolmat/benchmark/comparator.py [hunk 3]
```diff
@@ -118,14 +136,16 @@ def evaluate_errors_sample(
...
+                df_origin[subset], df_imputed[subset], df_mask[subset], metrics, on_mask
+            )
             list_errors.append(errors)
         df_errors = pd.DataFrame(list_errors)
         errors_mean = df_errors.mean(axis=0)
 
         return errors_mean
 
-    def compare(self, df: pd.DataFrame, verbose: bool = True):
+    def compare(self, df: pd.DataFrame, verbose: bool = True, metrics: Dict = {}, on_mask=True):
```
Comment (JulienRoussel77) [inline qolmat/benchmark/comparator.py:148]: Essaye de faire en sorte que la signature soit celle-là, pour simplifier l'utilisation :
def compare(self, df: pd.DataFrame, verbose: bool = True, metrics: List = ["mae", "wmape", "kl"]):

File: qolmat/benchmark/comparator.py [hunk 4]
```diff
@@ -60,30 +65,42 @@ def get_errors(
             dictionay of results obtained via different metrics
         """
 
+        if not on_mask:
+            df_mask = df_mask.replace(False, True)
```
[the remark this answers is not in the document]
Comment (JulienRoussel77) [reply qolmat/benchmark/comparator.py:69]: Concernant la structure de la fonction : définir un dictionaire de la forme :
dict_metrics = {"mae": utils.mean_absolute_error, ...}

puis faire :

for name_metric in list_metrics:
    dict_errors[name_metric] = dict_metrics[name_metric](df_origin, df_imputed, df_mask)

File: qolmat/tests/test_imputations/test_imputers.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+def test_ImputerMode_fit_transform(df: pd.DataFrame) -> None:
+    imputer = imputers.ImputerMode()
+    result = imputer.fit_transform(df)
+    expected = pd.DataFrame({"col1": [0, 0, 2, 3, 0], "col2": [-1, -1, 0.5, -1, 1.5]})
+    np.testing.assert_allclose(result, expected)
+
+
+@pytest.mark.parametrize("df", [pd.DataFrame({"col1": [1, 1, np.nan]})])
+def test_ImputerShuffle_fit_transform1(df: pd.DataFrame) -> None:
+    imputer = imputers.ImputerShuffle()
```
Comment (JulienRoussel77) [inline qolmat/tests/test_imputations/test_imputers.py:52]: C'est pas nécessaire de définir le random_state ici ?
